src/code/z_parameter.c

```diff
@@ -1145,26 +1149,484 @@ void Inventory_ChangeAmmo(s16 item, s16 ammoChange) {
     }
 }
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_parameter/Parameter_AddMagic.s")
+void Magic_Add(PlayState* play, s16 magicToAdd) {
+    if (((void)0, gSaveContext.save.playerData.magic) < ((void)0, gSaveContext.magicCapacity)) {
+        gSaveContext.magicToAdd += magicToAdd;
+        gSaveContext.isMagicRequested = true;
+    }
+}
+
+void Magic_Reset(PlayState* play) {
+    if ((gSaveContext.magicState != MAGIC_STATE_STEP_CAPACITY) && (gSaveContext.magicState != MAGIC_STATE_FILL)) {
+        sMagicMeterOutlinePrimRed = sMagicMeterOutlinePrimGreen = sMagicMeterOutlinePrimBlue = 255;
+        gSaveContext.magicState = MAGIC_STATE_IDLE;
+    }
+}
+
+/**
+ * Request to consume magic.
+ *
+ * @param magicToConsume the positive-valued amount to decrease magic by
+ * @param type how the magic is consumed.
+ * @return false if the request failed
+ */
+s32 Magic_Consume(PlayState* play, s16 magicToConsume, s16 type) {
+    InterfaceContext* interfaceCtx = &play->interfaceCtx;
+
+    // Magic is not acquired yet
+    if (!gSaveContext.save.playerData.isMagicAcquired) {
+        return false;
+    }
+
+    // Not enough magic available to consume
+    if ((gSaveContext.save.playerData.magic - magicToConsume) < 0) {
+        if (gSaveContext.magicCapacity != 0) {
+            play_sound(NA_SE_SY_ERROR);
+        }
+        return false;
+    }
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_parameter/func_80115D5C.s")
+    switch (type) {
+        case MAGIC_CONSUME_NOW:
+        case MAGIC_CONSUME_NOW_ALT:
+            // Consume magic immediately
+            // Ex. Deku Bubble
+            if ((gSaveContext.magicState == MAGIC_STATE_IDLE) ||
+                (gSaveContext.magicState == MAGIC_STATE_CONSUME_LENS)) {
+                if (gSaveContext.magicState == MAGIC_STATE_CONSUME_LENS) {
+                    play->actorCtx.lensActive = false;
+                }
+                if (gSaveContext.save.weekEventReg[14] & 8) {
+                    // Consumed Chateau Romani
```

Comment:
I know Anghelo has a PR out for doing some weekEventRegs defines, but it might make sense to make a define for this in this PR, since it pops up so often. I also have a minor wording quibble here; this whole function uses the word "consume" a ton, so I'd prefer "Drank Chateau Romani" or something just to make it clear that no, this is actually not related to magic consumption.

Reply:
Since there are many more parameter PRs coming, I think I'd prefer to name the weekeventreg once the Anghelo PR is in and avoid that conflict (I can do it nicely when I PR `Interface_Update` which has the remaining code on magic). I can change `consume` -> `drank` though (it's not my fault Link consumes a lot of various things on his adventures haha)